### DISCUSSION TITLE
stable-baselines DQN on highway env

### DISCUSSION THREAD

**Discussion by rjoseph24** [eleurent/highway-env#219] [Q&A]
Hi,

I have a question about training a DQN agent on the highway-env. The ego vehicle has to drive on the rightmost lane all the highest speed to maximize it's reward. In my case, it does understand to drive on the right most lane and I might add even at decent speed (not sure if it is the top speed of the vehicle). But it fails to avoid collisions with vehicles in the right lane. I might have trained the model for around a 100,000 timesteps by now.

Just want to make sure I am doing it the right way. Any help is appreciated.

**Comment by eleurent** [comment]
Hi Rohan,

It is generally difficult to debug a RL algorithm not converging to a good policy, there is no perfect answer to your question.
However, several factors can affect convergence:
- First, the hyperparameters have to be properly tuned (can you post which ones you used?)
- Second, the choice of observation type and corresponding network architecture matters a lot, and the default option is simple, but not very effective. Using more appropriate representations of the scene (like occupancy grids, images) or model architectures (transformers...). (you should still be able to obtain okay-ish results and consistent collision avoidance with the default configuration though)
- I think that the last time I trained a model on highway, i used 2e5 timesteps, which may make a difference. 

To make the training faster, you may want to configure the environment to have fewer vehicles (and shorter episodes).

**Reply by rjoseph24** [reply]
Hi Edouard,

Thank you for your response.
A snippet of the code I am running for training:

env = gym.make("highway-v0")
model = DQN("MlpPolicy", env, verbose=1)
model.learn(total_timesteps=30000) # might have run those 5 times so far

I am not tuning any hyperparameters, just going with the default values.
As recommended by you, maybe reaching 2e5 should allow the ego vehicle to avoid collisions as well.
Also, for faster training as recommended by you, I will try to change the config of the env.

Do provide any other suggestions if applicable.

Thanks!

**Reply by rjoseph24** [reply]
Also, using the DQN of: from stable_baselines3 import DQN

Have also used A2C and PPO on the same environment with no luck on the performance so far.

**Reply by eleurent** [reply]
I have tried DQN and PPO, and both worked fine.
I recommend that you use a smaller discount factor (gamma) than the default value (which is probably 0.95, i.e. horizon of about 40 steps). Since the action duration is quite high (1s), it is sufficient to only optimize a horizon about 10 steps (10 seconds) in the future, i.e. you can use gamma = 0.8

**Reply by eleurent** [reply]
In fact, I recommend that you try the hyperparameters that I used in https://github.com/eleurent/highway-env/blob/master/scripts/stablebaselines_highway.py

**Reply by rjoseph24** [reply]
Thank you for your reply Edouard. I will try to run your code as is and see if I get the results.

**Reply by rjoseph24** [reply]
Hi Edouard,

Once again, thank you for your assistance. With your suggestions the models seem to be performing much better. However, is there a documentation page which clears states the different configurations that can be changed in the environment and the hyperparameters of the model elaborated further. This will be helpful for me who has very little understanding of reinforcement learning and the algorithms.

Any help is greatly appreciated!

**Reply by eleurent** [reply]
> is there a documentation page which clears states the different configurations that can be changed in the environment 

You can find more information about the different environments and there configurations on this page: https://highway-env.readthedocs.io/en/latest/quickstart.html#all-the-environments
Most of these configurations should be self-explanatory, but some of them may require you to have a look in the source code to see how they are used.

> ... and the hyperparameters of the model elaborated further.

In principle, the role of this library is only to describe the environment, while the agent/RL algorithm solving the task is out of its scope and should be handled separately. Moreover, there is not one single right answer to this question, but many possible algorithms/implementations/tunings.
However, I still wanted to provide a few examples of agents solving the environments using well-known RL libraries, while giving the complete list of hyperparameters. It was the goal of scripts such as this one: https://github.com/eleurent/highway-env/blob/master/scripts/stablebaselines_highway.py

But I agree that this format is not very informative, since it does not explain why/how these hyperparameters were chosen.
I may add a page addressing this question in the documentation.
